src/app/app.component.ts

```diff
@@ -0,0 +1,100 @@
+  searchTermByCharacters = new Subject<string>();
+  charactersResults$: Observable<any>;
+  planetAndCharactersResults$: Observable<any>;
+  isLoading: boolean = false;
+
+  constructor(private mockDataService: MockDataService) {}
+
+  ngOnInit(): void {
+    this.initLoadingState();
+    this.initCharacterEvents();
+  }
+
+  changeCharactersInput(element): void {
+    // 1.1. Add functionality to changeCharactersInput method. Changes searchTermByCharacters Subject value on input change.
+
+    const inputValue: string = element.target.value;
+    this.searchTermByCharacters.next(inputValue);
+  }
+
+  initCharacterEvents(): void {
+    // 1.2. Add API call on each user input. Use mockDataService.getCharacters - to make get request.
+
+    // 2. Since we don't want to spam our service add filter by input value and do not call API until a user enters at least 3 chars.
+
+    // 3. Add debounce to prevent API calls until user stop typing.
+
+    this.charactersResults$ = this.searchTermByCharacters
+        .pipe(
+            map(characters => (characters ? characters.trim() : "")),
+            filter(inputValue => inputValue.length >= 3),
+            debounceTime(500),
+            switchMap(inputValue => this.mockDataService.getCharacters(inputValue)),
+        );
+  }
+
+  loadCharactersAndPlanet(): void {
+    // 4. On clicking the button 'Load Characters And Planets', it is necessary to process two requests and combine the results of both requests into one result array. As a result, a list with the names of the characters and the names of the planets is displayed on the screen.
+    // Your code should look like this: this.planetAndCharactersResults$ = /* Your code */
+
+    this.planetAndCharactersResults$ = forkJoin([
+      this.mockDataService.getCharacters(),
+      this.mockDataService.getPlatents()
+    ]).pipe(
+        map(([characters, planets]) => {
+          return [...characters, ...planets].map(item => item);
+        })
+    );
+  }
+
+  initLoadingState(): void {
+    /* 5.1. Let's add loader logic to our page. For each request, we have an observable that contains the state of the request. When we send a request the value is true, when the request is completed, the value becomes false. You can get value data with mockDataService.getCharactersLoader() and mockDataService.getPlanetLoader().
+
+    - Combine the value of each of the streams.
+    - Subscribe to changes
+    - Check the received value using the areAllValuesTrue function and pass them to the isLoading variable. */
+
+    combineLatest([
+      this.mockDataService.getCharactersLoader(),
+      this.mockDataService.getPlanetLoader()
+    ]).subscribe(([charactersLoading, planetsLoading]) => {
+      this.isLoading = this.areAllValuesTrue([charactersLoading, planetsLoading]);
+    });
+  }
+
+  ngOnDestroy(): void {
+    // 5.2 Unsubscribe from all subscriptions
+    const subscriptions = [];
+    subscriptions.push(
+        this.charactersResults$,
```

Comment:
I don't see charactersResults$ is a subscription. It's rather observable.
Same with second param. Corect me if I'm wrong

Reply:
that's what I was confused about, I see that here is subscription but don't get how to unsubscribe that 

<img width="685" alt="image" src="https://user-images.githubusercontent.com/36797396/216930384-1679b536-45ef-4aac-9314-e2a52f1fd41c.png">

Reply:
first, I mentioned subscribing can be avoided above. But if you have to, then put this subscription into a separate class property.

```suggestion
        readonly subscription: Subscription;
        
        someMethod() {
          this.subscription = combineLatest(...).subscribe(...)
        }
        
        ngOnDestroy() {
          this.subscription.unsubscribe()
        }
```

Reply:
done